### PR TITLE
Introduce a xcframework dependency type

### PR DESCRIPTION
**Pull request by lakpa**
### Short description 📝

This PR introduces a new dependency type to support `xcframework`. 

### Solution 📦

A graph node to represent xcframework dependency called `XCFrameworkNode` is introduced which inherits from `PrecompiledNode`, the `XCframeworkNode` can leverage existing lookup however the only exception we need to consider is during `embedding phase` when we need to specifically look for `xcframework` file type to skip getting the xcframework being picked up as `precompiled` to skip `embed script build phase`  and embed the respective file reference with required `CodeSignOnCopy` attribute instead.

### CODE REVIEW

File: fixtures/ios_app_with_xcframeworks/Tests/xcframeworksTests.swift [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+import XCTest
+
+@testable import xcframeworks
```
Comment (kwridan) [inline fixtures/ios_app_with_xcframeworks/Tests/xcframeworksTests.swift]: typo: `@testable import App`

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift [hunk 2]
```diff
@@ -88,6 +88,46 @@ final class LinkGeneratorErrorTests: XCTestCase {
...
+        let fileRelativePath = RelativePath("./Test.xcframework")
+        let fileElements = ProjectFileElements()
+        fileElements.addFileElement(from: from,
+                                          fileAbsolutePath: fileAbsolutePath,
+                                          fileRelativePath: fileRelativePath,
+                                          name: nil,
+                                          toGroup: group,
+                                          pbxproj: pbxproj)
+
+        let wakaFile = fileElements.file(path: fileAbsolutePath)
```
Comment (kwridan) [inline Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: I wonder if we can benefit from a test helper here to make the test more concise.

```swift
// Given
let xcframeworkPath = AbsolutePath("/Frameworks/Test.xcframework")
let dependencies: [DependencyReference] = [
  .absolute(xcframeworkPath)
]
let fileElements = createFileElements(for: xcframeworkPath)
// ...

```

Note: "waka" was the name of a framework used in another test, there is no need to use it here as the framework we have here is called "test", the name doesn't really impact the test.

File: Tests/TuistCoreTests/Graph/Nodes/TargetNodeTests.swift [hunk 3]
```diff
@@ -49,14 +49,15 @@ final class TargetNodeTests: XCTestCase {
         XCTAssertNotEqual(c1, d)
     }
 
-    func test_encode() {
+    func test_encode() throws {
         // Given
         let library = LibraryNode.test()
         let framework = FrameworkNode.test()
         let cocoapods = CocoaPodsNode.test()
+        let xcframework = try XCFrameworkNode.test()
```
Comment (kwridan) [inline Tests/TuistCoreTests/Graph/Nodes/TargetNodeTests.swift]: Does this helper exist? is there a reason it's throwing?

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift [hunk 4]
```diff
@@ -88,6 +88,46 @@ final class LinkGeneratorErrorTests: XCTestCase {
...
+                                       fileElements: fileElements,
+                                       sourceRootPath: sourceRootPath)
+
+        // Then
+        let copyBuildPhase: PBXCopyFilesBuildPhase? = pbxTarget.buildPhases.last as? PBXCopyFilesBuildPhase
+        XCTAssertEqual(copyBuildPhase?.name, "Embed Frameworks")
+        let wakaBuildFile: PBXBuildFile? = copyBuildPhase?.files?.last
+        XCTAssertEqual(wakaBuildFile?.file, wakaFile)
+        let settings: [String: [String]]? = wakaBuildFile?.settings as? [String: [String]]
+        XCTAssertEqual(settings, ["ATTRIBUTES": ["CodeSignOnCopy"]])
```
Comment (kwridan) [inline Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: Similarly here we could benefit from some array comparisons to make the asserts more readable 

_(this is somewhat subjective)_
```swift
       // Then
        let copyBuildPhase = try XCTUnwrap(pbxTarget.embedFrameworksBuildPhases().first)
        let buildFiles = try XCTUnwrap(copyBuildPhase.files)
        XCTAssertEqual(buildFiles.map { $0.file?.path }, [
            "Test.xcframework"
        ])
        XCTAssertEqual(buildFiles.map { $0.settings as? [String: [String]] }, [
            ["ATTRIBUTES": ["CodeSignOnCopy"]]
        ])
```

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift (hunk 2 above)
Comment (lakpa) [reply Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: Agreed, we could make code concise by using a helper.

File: Tests/TuistCoreTests/Graph/Nodes/TargetNodeTests.swift (hunk 3 above)
Comment (lakpa) [reply Tests/TuistCoreTests/Graph/Nodes/TargetNodeTests.swift]: Thanks for spotting, i was running some trials, it got left over after. Will remove it.

File: Sources/TuistGenerator/Generator/LinkGenerator.swift [hunk 5]
```diff
@@ -169,7 +169,19 @@ final class LinkGenerator: LinkGenerating {
 
         try dependencies.forEach { dependency in
             if case let GraphDependencyReference.absolute(path) = dependency {
-                precompiledFrameworkPaths.append(path)
+                if path.extension == "xcframework" {
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/LinkGenerator.swift]: 👏

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift [hunk 6]
```diff
@@ -88,6 +88,39 @@ final class LinkGeneratorErrorTests: XCTestCase {
             XCTAssertEqual($0 as? LinkGeneratorError, LinkGeneratorError.missingProduct(name: "Test"))
         }
     }
+    
+    func test_generateEmbedPhase_setupEmbedFrameworksBuildPhase_whenXCFrameworkIsPresent() throws {
```
Comment (pepibumur) [inline Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: Beautifully tested 👏

File: features/generate.feature [hunk 7]
```diff
@@ -230,3 +230,11 @@ Scenario: The project is an iOS application with watch app (ios_app_with_watchap
     Then I should be able to build for watchOS the scheme App
     Then the product 'App.app' with destination 'Debug-iphoneos' contains resource 'Watch/WatchApp.app'
     Then the product 'WatchApp.app' with destination 'Debug-watchos' contains extension 'WatchAppExtension' 
+
+Scenario: The project is an iOS application with Xcframeworks (ios_app_with_xcframeworks)
```
Comment (pepibumur) [inline features/generate.feature]: Nitpick: Xcframeworks -> xcframeworks

**Comment by lakpa** [comment]
> Hey @lakpa! Just wanted to say you did an amazing job here 👏. I'm very excited about seeing xcframeworks' support landing in the project.

Could manage to do with guidance from you and @kwridan , thanks for your guidance!!!

File: fixtures/ios_app_with_xcframeworks/Tests/xcframeworksTests.swift (hunk 1 above)
Comment (lakpa) [reply fixtures/ios_app_with_xcframeworks/Tests/xcframeworksTests.swift]: wondering if I could include the static lib fixture in next PR as follow up work is needed to support static libs, which hasn't accounted in this PR in regards to `binary path`.

**Comment by lakpa** [comment]
> * Worth adding a static library xcframework fixture too

wondering if I could include the static lib fixture in next PR as follow up work is needed to support static libs, which hasn't accounted in this PR in regards to `binary path`.

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift [hunk 8]
```diff
@@ -482,4 +515,19 @@ final class LinkGeneratorErrorTests: XCTestCase {
 
         return projectFileElements
     }
+    
+    private func createFileElements(fileAbsolutePath: AbsolutePath,
```
Comment (kwridan) [inline Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: awesome! thanks for making this helper 👍 

One last suggestion to simplify it even further, we can prime the `ProjectFileElements` directly with the fileElement.

```swift
 private func createFileElements(fileAbsolutePath: AbsolutePath) -> ProjectFileElements {
        let fileElements = ProjectFileElements()
        fileElements.elements[fileAbsolutePath] = PBXFileReference(path: fileAbsolutePath.basename)
        return fileElements
    }
```

Something like this makes it an even more re-usable helper here as it not specific to the `Test.xcframework`

File: Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift [hunk 9]
```diff
@@ -88,6 +88,39 @@ final class LinkGeneratorErrorTests: XCTestCase {
...
+
+        // Then
+        let copyBuildPhase = try XCTUnwrap(pbxTarget.embedFrameworksBuildPhases().first)
+        XCTAssertEqual(copyBuildPhase.name, "Embed Frameworks")
+        let buildFiles = try XCTUnwrap(copyBuildPhase.files)
+        XCTAssertEqual(buildFiles.map { $0.file?.path }, [ "Test.xcframework"])
+        XCTAssertEqual(buildFiles.map { $0.settings as? [String: [String]] }, [
+            ["ATTRIBUTES": ["CodeSignOnCopy"]]
+        ])
+    }
```
Comment (kwridan) [inline Tests/TuistGeneratorTests/Generator/LinkGeneratorTests.swift]: 👍

**Pull request merged**